### PR TITLE
local_segment in segmented_iterator_traits

### PR DESCRIPTION
**Pull request by atrantan**
This patch tries to give some means to get the closest _local_segment_ of a _partitioned_vector_ from a _segment_. Since `local_segment_vector_iterator` and `segment_vector_iterator` are both designed from the same base iterator type, my solution is to grab the base_iterator of the **input** segment_vector_iterator (via conversion functions) and use this to retrieve the **output** local_segment_vector_iterator

### CODE REVIEW

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp [hunk 1]
```diff
@@ -555,6 +560,11 @@ namespace hpx
           : base_type(it), data_(data)
         {}
 
+        operator BaseIter const &() const
```
Comment (hkaiser) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: Shouldn't this be `explicit`?

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp [hunk 2]
```diff
@@ -521,6 +521,11 @@ namespace hpx
...
```
Comment (hkaiser) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: Shouldn't this be `explicit`?

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp [hunk 3]
```diff
@@ -852,6 +862,14 @@ namespace hpx { namespace traits
                 iter.get_global_index());
         }
 
+       //  Conceptually this function is supposed to retrieve the closest
+       //  local segment of the segment i in locality id.
+        static local_segment_iterator local_segment(
```
Comment (hkaiser) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: The indentation is somehow off here.

**Comment by hkaiser** [comment]
Could you give a use case please explaining why those changes are necessary?

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp (hunk 2 above)
Comment (atrantan) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: The template parameter `BaseIter` of `segment_vector_iterator` seems to be private and `segmented_iterator_traits` seems to not contain any `typedef` corresponding to that type.

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp (hunk 1 above)
Comment (atrantan) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: Same problem

**Comment by atrantan** [comment]
Example: 

1. Partitions spread across localities. 
2.Team of threads in a specific locality want to work on a subrange of those partitions 
3. Need to filter the local partitions among those partitions and  say : local_partition(0) for thread 0, local_partition(1) for thread 1, etc.

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: I meant, shouldn't this be:

    explicit operator BaseIter const& () { ... }

?

**Comment by hkaiser** [comment]
Isn't this functionality already available through the functions `segmented_iterator_traits<>::local()` and `partiioned_vector<>::segment_begin()`/`partitioned_vector<>::segment_end()`?

**Comment by atrantan** [comment]
`segmented_iterator_traits<>::local()` seems to return a `local_iterator` but i need a `local_segment_iterator`. I'm effectively using `partitioned_vector<>::segment_begin()`/`partitioned_vector<>::segment_end()` to iterate over partitions.

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp (hunk 2 above)
Comment (atrantan) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: `segmented_iterator_traits` needs to know the template parameter `BaseIter` of `segment_vector_iterator` to _explicitly_ cast the input segment_vector_iterator inside the newly implemented `local_segment()` method. But there's neither public `segment_vector_iterator`'s inner typedef  (think it was intentional), nor `segmented_iterator_traits`'s inner typedef corresponding to this `BaseIter`

File: hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp (hunk 3 above)
Comment (atrantan) [inline hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: Done in 4db6771f3eae2524b02d7cf5c68ad03ef1757e49

**Comment by atrantan** [comment]
I'm closing this PR because of uncertainty about enhancing the functionality of `segmented_iterator_traits`

**Pull request closed without merging**